In [16]:
import numpy as np

In [19]:
# jacobian
def jacobi(A, b, x0, tol, max_iterations):
    n = len(b)
    x = x0.copy()
    for k in range(max_iterations):
        x_new = np.zeros_like(x)
        for i in range(n):
            s = sum(A[i][j] * x[j] for j in range(n) if j != i)
            x_new[i] = (b[i] - s) / A[i][i]
        # Check for convergence
        if np.linalg.norm(x_new - x, ord=np.inf) < tol:
            return x_new, k
        x = x_new
    return x, max_iterations


In [20]:
A = np.array([[10,-1,2,0,0],[-1,11,-1,3,0],[2,-1,10,-1,0],[0,3,-1,8,-2],[0,0,0,-2,9]])
b = np.array([14,30,26,25,37])
A = A.astype(float)
b = b.astype(float)
x0 = np.zeros_like(b)
tol = 1*(10**-6)
max_iterations = 100

In [32]:
%%time
solution, iterations = jacobi(A, b, x0, tol, max_iterations)
print(f"Solution: {solution}")
print(f"Iterations: {iterations}")

Solution: [0.99999995 2.00000025 2.99999986 4.00000008 4.99999983]
Iterations: 18
CPU times: user 3.22 ms, sys: 0 ns, total: 3.22 ms
Wall time: 2.84 ms


Jacobian method
- took 18 iterations to converge to a solution
- time to run: 2.98 ms

In [13]:
# gauss-seidel
def gauss_seidel(A, b, x0, tol, max_iterations):
    n = len(b)
    x = x0.copy()
    for k in range(max_iterations):
        x_new = x.copy()
        for i in range(n):
            s1 = sum(A[i][j] * x_new[j] for j in range(i)) # Using already updated values
            s2 = sum(A[i][j] * x[j] for j in range(i + 1, n)) # Using old values
            x_new[i] = (b[i] - s1 - s2) / A[i][i]
            # Check for convergence
        if np.linalg.norm(x_new - x, ord=np.inf) < tol:
            return x_new, k
        x = x_new
    return x, max_iterations

In [15]:
A = np.array([[10,-1,2,0,0],[-1,11,-1,3,0],[2,-1,10,-1,0],[0,3,-1,8,-2],[0,0,0,-2,9]])
b = np.array([14,30,26,25,37])
A = A.astype(float)
b = b.astype(float)
x0 = np.zeros_like(b)
tol = 1*(10**-6)
max_iterations = 100

Solution: [1.00000003 2.00000002 2.99999999 3.99999998 5.        ]
Iterations: 10


In [22]:
%%time
solution, iterations = gauss_seidel(A, b, x0, tol, max_iterations)
print(f"Solution: {solution}")
print(f"Iterations: {iterations}")

Solution: [1.00000003 2.00000002 2.99999999 3.99999998 5.        ]
Iterations: 10
CPU times: user 2.24 ms, sys: 882 µs, total: 3.12 ms
Wall time: 2.68 ms


Gauss-Seidel method
- took 10 iterations to converge to a solution
- time to run: 2.68 ms